src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIMarsConnection.cs

```diff
@@ -0,0 +1,314 @@
+        {
+            foreach (SNIMarsHandle handle in _sessions.Values)
+            {
+                handle.HandleReceiveError();
+            }
+        }
+
+        /// <summary>
+        /// Process a send completion
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="sniErrorCode">SNI error code</param>
+        public void HandleSendComplete(SNIPacket packet, uint sniErrorCode)
+        {
+            packet.InvokeCompletionCallback(sniErrorCode);
+        }
+
+        /// <summary>
+        /// Process a receive completion
+        /// </summary>
+        /// <param name="packet">SNI packet</param>
+        /// <param name="sniErrorCode">SNI error code</param>
+        public void HandleReceiveComplete(SNIPacket packet, uint sniErrorCode)
+        {
+            SNISMUXHeader currentHeader = null;
+            SNIPacket currentPacket = null;
+            SNIMarsHandle currentSession = null;
+
+
+            if (sniErrorCode != TdsEnums.SNI_SUCCESS)
+            {
+                lock (this)
+                {
+                    HandleReceiveError();
+                    return;
+                }
+            }
+
+            while (true)
+            {
+                lock (this)
+                {
+                    if (_currentHeaderByteCount != SNISMUXHeader.HEADER_LENGTH)
+                    {
+                        currentHeader = null;
+                        currentPacket = null;
+                        currentSession = null;
+
+                        while (_currentHeaderByteCount != SNISMUXHeader.HEADER_LENGTH)
+                        {
+                            int bytesTaken = packet.TakeData(_headerBytes, _currentHeaderByteCount, SNISMUXHeader.HEADER_LENGTH - _currentHeaderByteCount);
+                            _currentHeaderByteCount += bytesTaken;
+
+                            if (bytesTaken == 0)
+                            {
+                                sniErrorCode = ReceiveAsync(ref packet);
+
+                                if (sniErrorCode == TdsEnums.SNI_SUCCESS_IO_PENDING)
+                                {
+                                    return;
+                                }
+
+                                HandleReceiveError();
+                                return;
+                            }
+                        }
+
+                        _currentHeader = new SNISMUXHeader();
+                        _currentHeader.SMID = _headerBytes[0];
+                        _currentHeader.flags = _headerBytes[1];
+                        _currentHeader.sessionId = BitConverter.ToUInt16(_headerBytes, 2);
+                        _currentHeader.length = BitConverter.ToUInt32(_headerBytes, 4) - SNISMUXHeader.HEADER_LENGTH;
+                        _currentHeader.sequenceNumber = BitConverter.ToUInt32(_headerBytes, 8);
+                        _currentHeader.highwater = BitConverter.ToUInt32(_headerBytes, 12);
```

Comment:
Just as a style thing, consider using object initialization syntax to highlight the configuration being done here, e.g.
```C#
_currentHeader = new SNISMUXHeader()
{
    SMID = _headerBytes[0],
    flags = _headerBytes[1],
    sessionId = BitConverter.ToUInt16(_headerBytes, 2);
    length = BitConverter.ToUInt32(_headerBytes, 4) - SNISMUXHeader.HEADER_LENGTH,
    sequenceNumber = BitConverter.ToUInt32(_headerBytes, 8),
    highwater = BitConverter.ToUInt32(_headerBytes, 12)
};